pkg/manager/member/dm_master_member_manager.go

```diff
@@ -0,0 +1,775 @@
+		force := NeedForceUpgrade(dc.Annotations)
+		if force {
+			dc.Status.Master.Phase = v1alpha1.UpgradePhase
+			setUpgradePartition(newMasterSet, 0)
+			errSTS := updateStatefulSet(mmm.setControl, dc, newMasterSet, oldMasterSet)
+			return controller.RequeueErrorf("dmcluster: [%s/%s]'s dm-master needs force upgrade, %v", ns, dcName, errSTS)
+		}
+	}
+
+	// TODO: dm add rolling update later
+	// if !templateEqual(newMasterSet, oldMasterSet) || dc.Status.Master.Phase == v1alpha1.UpgradePhase {
+	//	if err := mmm.masterUpgrader.Upgrade(dc, oldMasterSet, newMasterSet); err != nil {
+	//		return err
+	//	}
+	// }
+
+	// TODO: dm add scaler
+	//if err := mmm.masterScaler.Scale(dc, oldMasterSet, newMasterSet); err != nil {
+	//	return err
+	//}
+
+	// TODO: dm add auto failover
+	// if mmm.autoFailover {
+	//	if mmm.shouldRecover(tc) {
+	//		mmm.masterFailover.Recover(tc)
+	//	} else if dc.MasterAllPodsStarted() && !dc.MasterAllMembersReady() || tc.MasterAutoFailovering() {
+	//		if err := mmm.masterFailover.Failover(tc); err != nil {
+	//			return err
+	//		}
+	//	}
+	// }
+
+	return updateStatefulSet(mmm.setControl, dc, newMasterSet, oldMasterSet)
+}
+
+func (mmm *masterMemberManager) syncDMClusterStatus(dc *v1alpha1.DMCluster, set *apps.StatefulSet) error {
+	if set == nil {
+		// skip if not created yet
+		return nil
+	}
+
+	ns := dc.GetNamespace()
+	dcName := dc.GetName()
+
+	dc.Status.Master.StatefulSet = &set.Status
+
+	upgrading, err := mmm.masterStatefulSetIsUpgrading(set, dc)
+	if err != nil {
+		return err
+	}
+	if upgrading {
+		dc.Status.Master.Phase = v1alpha1.UpgradePhase
+	} else if dc.MasterStsDesiredReplicas() != *set.Spec.Replicas {
+		dc.Status.Master.Phase = v1alpha1.ScalePhase
+	} else {
+		dc.Status.Master.Phase = v1alpha1.NormalPhase
+	}
+
+	dmClient := controller.GetMasterClient(mmm.masterControl, dc)
+
+	mastersInfo, err := dmClient.GetMasters()
+	if err != nil {
+		dc.Status.Master.Synced = false
+		// get endpoints info
+		eps, epErr := mmm.epsLister.Endpoints(ns).Get(controller.DMMasterMemberName(dcName))
+		if epErr != nil {
+			return fmt.Errorf("syncDMClusterStatus: failed to get endpoints %s for cluster %s/%s, err: %s, epErr %s", controller.DMMasterMemberName(dcName), ns, dcName, err, epErr)
+		}
+		// dm-master service has no endpoints
+		if eps != nil && len(eps.Subsets) == 0 {
+			return fmt.Errorf("%s, service %s/%s has no endpoints", err, ns, controller.DMMasterMemberName(dcName))
+		}
+		return err
+	}
+
+	leader, err := dmClient.GetLeader()
+	if err != nil {
+		dc.Status.Master.Synced = false
+		return err
+	}
+	masterStatus := map[string]v1alpha1.MasterMember{}
+	for _, master := range mastersInfo {
+		id := master.MemberID
+		var clientURL string
+		if len(master.ClientURLs) > 0 {
+			clientURL = master.ClientURLs[0]
+		}
+		name := master.Name
+		if len(name) == 0 {
+			klog.Warningf("dm-master member: [%s] doesn't have a name, and can't get it from clientUrls: [%s], dm-master Info: [%v] in [%s/%s]",
```

Comment:
Why add `can't get it from clientUrls` this description, you do not try to get it from clientUrls.